File: backend/app/telemetry.py
"""
API usage telemetry — two complementary classes:

  CostLedger   — persistent, thread-safe cumulative cost tracker (survives restarts)
  ActivityLog  — ephemeral in-memory ring buffer for the live UI activity feed
"""

import json
import logging
import threading
from datetime import datetime, timezone

from .config import (
    EXTRACTION_MODEL,
    EMBEDDING_MODEL,
    XAI_PRICE_INPUT_PER_M,
    XAI_PRICE_CACHED_PER_M,
    XAI_PRICE_OUTPUT_PER_M,
    GOOGLE_EXTRACTION_MODEL,
    GOOGLE_PRICE_INPUT_PER_M,
    GOOGLE_PRICE_OUTPUT_PER_M,
    ANTHROPIC_EXTRACTION_MODEL,
    ANTHROPIC_PRICE_INPUT_PER_M,
    ANTHROPIC_PRICE_OUTPUT_PER_M,
    VOYAGE_PRICE_PER_M,
)

logger = logging.getLogger("memory-server")


# ── CostLedger ─────────────────────────────────────────────────────────────────


class CostLedger:
    """Thread-safe accumulator for API token usage and USD cost.

    Persists to DATA_DIR/costs.json so totals survive container restarts.
    """

    _TEMPLATE = {
        "xai": {
            "calls": 0,
            "prompt_tokens": 0,
            "cached_tokens": 0,
            "completion_tokens": 0,
            "cost_usd": 0.0,
        },
        "google": {
            "calls": 0,
            "prompt_tokens": 0,
            "completion_tokens": 0,
            "cost_usd": 0.0,
        },
        "anthropic": {
            "calls": 0,
            "prompt_tokens": 0,
            "completion_tokens": 0,
            "cost_usd": 0.0,
        },
        "voyage": {
            "calls": 0,
            "tokens": 0,
            "cost_usd": 0.0,
        },
        "total_cost_usd": 0.0,
        "last_updated": "",
    }

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._data: dict = {}
        self._path: str = ""

    def init(self, data_dir: str) -> None:
        self._path = f"{data_dir}/costs.json"
        try:
            with open(self._path) as f:
                self._data = json.load(f)
            logger.info("Cost ledger loaded from %s", self._path)
        except FileNotFoundError:
            self._data = json.loads(json.dumps(self._TEMPLATE))
            logger.info("Cost ledger initialised (new)")
        except Exception as e:
            logger.warning("Cost ledger load error: %s — starting fresh", e)
            self._data = json.loads(json.dumps(self._TEMPLATE))

    def _save(self) -> None:
        """Write to disk. Must be called while holding _lock."""
        if not self._path:
            return
        try:
            with open(self._path, "w") as f:
                json.dump(self._data, f, indent=2)
        except Exception as e:
            logger.warning("Cost ledger save error: %s", e)

    def _update_total(self) -> None:
        """Recompute total_cost_usd from all provider buckets. Caller must hold _lock."""
        self._data["total_cost_usd"] = round(
            self._data["xai"]["cost_usd"]
            + self._data.get("google", {}).get("cost_usd", 0.0)
            + self._data.get("anthropic", {}).get("cost_usd", 0.0)
            + self._data["voyage"]["cost_usd"],
            8,
        )
        self._data["last_updated"] = datetime.now(timezone.utc).isoformat()

    def record_xai(self, prompt_tokens: int, cached_tokens: int, completion_tokens: int) -> None:
        cost = (
            (prompt_tokens - cached_tokens) * XAI_PRICE_INPUT_PER_M / 1_000_000
            + cached_tokens * XAI_PRICE_CACHED_PER_M / 1_000_000
            + completion_tokens * XAI_PRICE_OUTPUT_PER_M / 1_000_000
        )
        with self._lock:
            x = self._data["xai"]
            x["calls"] += 1
            x["prompt_tokens"] += prompt_tokens
            x["cached_tokens"] += cached_tokens
            x["completion_tokens"] += completion_tokens
            x["cost_usd"] = round(x["cost_usd"] + cost, 8)
            self._update_total()
            self._save()

    def record_google(self, prompt_tokens: int, completion_tokens: int) -> None:
        cost = (
            prompt_tokens * GOOGLE_PRICE_INPUT_PER_M / 1_000_000
            + completion_tokens * GOOGLE_PRICE_OUTPUT_PER_M / 1_000_000
        )
        with self._lock:
            # Backfill "google" key for ledgers created before this provider existed
            if "google" not in self._data:
                self._data["google"] = {"calls": 0, "prompt_tokens": 0, "completion_tokens": 0, "cost_usd": 0.0}
            g = self._data["google"]
            g["calls"] += 1
            g["prompt_tokens"] += prompt_tokens
            g["completion_tokens"] += completion_tokens
            g["cost_usd"] = round(g["cost_usd"] + cost, 8)
            self._update_total()
            self._save()

    def record_anthropic(self, prompt_tokens: int, completion_tokens: int) -> None:
        cost = (
            prompt_tokens * ANTHROPIC_PRICE_INPUT_PER_M / 1_000_000
            + completion_tokens * ANTHROPIC_PRICE_OUTPUT_PER_M / 1_000_000
        )
        with self._lock:
            # Backfill "anthropic" key for ledgers created before this provider existed
            if "anthropic" not in self._data:
                self._data["anthropic"] = {"calls": 0, "prompt_tokens": 0, "completion_tokens": 0, "cost_usd": 0.0}
            a = self._data["anthropic"]
            a["calls"] += 1
            a["prompt_tokens"] += prompt_tokens
            a["completion_tokens"] += completion_tokens
            a["cost_usd"] = round(a["cost_usd"] + cost, 8)
            self._update_total()
            self._save()

    def record_voyage(self, tokens: int) -> None:
        cost = tokens * VOYAGE_PRICE_PER_M / 1_000_000
        with self._lock:
            v = self._data["voyage"]
            v["calls"] += 1
            v["tokens"] += tokens
            v["cost_usd"] = round(v["cost_usd"] + cost, 8)
            self._update_total()
            self._save()

    def snapshot(self) -> dict:
        with self._lock:
            return json.loads(json.dumps(self._data))

    def reset(self) -> None:
        with self._lock:
            self._data = json.loads(json.dumps(self._TEMPLATE))
            self._data["last_updated"] = datetime.now(timezone.utc).isoformat()
            self._save()


# ── ActivityLog ────────────────────────────────────────────────────────────────


class ActivityLog:
    """In-memory ring buffer of the last N API calls (xAI + Voyage).

    Not persisted — resets on container restart. Intended for the live
    activity feed in the web UI so you can confirm extractions are happening.
    """

    MAX_ENTRIES = 200

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._entries: list[dict] = []

    def _append(self, entry: dict) -> None:
        with self._lock:
            self._entries.append(entry)
            if len(self._entries) > self.MAX_ENTRIES:
                self._entries = self._entries[-self.MAX_ENTRIES:]

    def record_xai(
        self,
        prompt_tokens: int,
        cached_tokens: int,
        completion_tokens: int,
        cost_usd: float,
        operation: str = "extraction",
    ) -> None:
        self._append({
            "ts": datetime.now(timezone.utc).isoformat(),
            "api": "xai",
            "model": EXTRACTION_MODEL,
            "operation": operation,
            "prompt_tokens": prompt_tokens,
            "cached_tokens": cached_tokens,
            "completion_tokens": completion_tokens,
            "tokens": None,
            "cost_usd": round(cost_usd, 8),
        })

    def record_google(
        self,
        prompt_tokens: int,
        completion_tokens: int,
        cost_usd: float,
        operation: str = "extraction",
    ) -> None:
        self._append({
            "ts": datetime.now(timezone.utc).isoformat(),
            "api": "google",
            "model": GOOGLE_EXTRACTION_MODEL,
            "operation": operation,
            "prompt_tokens": prompt_tokens,
            "cached_tokens": None,
            "completion_tokens": completion_tokens,
            "tokens": None,
            "cost_usd": round(cost_usd, 8),
        })

    def record_anthropic(
        self,
        prompt_tokens: int,
        completion_tokens: int,
        cost_usd: float,
        operation: str = "extraction",
    ) -> None:
        self._append({
            "ts": datetime.now(timezone.utc).isoformat(),
            "api": "anthropic",
            "model": ANTHROPIC_EXTRACTION_MODEL,
            "operation": operation,
            "prompt_tokens": prompt_tokens,
            "cached_tokens": None,
            "completion_tokens": completion_tokens,
            "tokens": None,
            "cost_usd": round(cost_usd, 8),
        })

    def record_voyage(self, tokens: int, cost_usd: float) -> None:
        self._append({
            "ts": datetime.now(timezone.utc).isoformat(),
            "api": "voyage",
            "model": EMBEDDING_MODEL,
            "operation": "embed",
            "prompt_tokens": None,
            "cached_tokens": None,
            "completion_tokens": None,
            "tokens": tokens,
            "cost_usd": round(cost_usd, 8),
        })

    def recent(self, limit: int = 50) -> list[dict]:
        with self._lock:
            return list(reversed(self._entries[-limit:]))


# ── Module-level singletons ────────────────────────────────────────────────────

ledger = CostLedger()
activity_log = ActivityLog()

File: plugin/src/telemetry.ts
/**
 * API usage telemetry — two complementary classes:
 *
 *   CostLedger   — persistent cumulative cost tracker (survives restarts)
 *   ActivityLog  — in-memory ring buffer for activity feed
 */

import {
	ACTIVITY_MAX,
	ANTHROPIC_EXTRACTION_MODEL,
	ANTHROPIC_PRICE_INPUT_PER_M,
	ANTHROPIC_PRICE_OUTPUT_PER_M,
	DATA_DIR,
	EMBEDDING_MODEL,
	EXTRACTION_MODEL,
	GOOGLE_EXTRACTION_MODEL,
	GOOGLE_PRICE_INPUT_PER_M,
	GOOGLE_PRICE_OUTPUT_PER_M,
	VOYAGE_PRICE_PER_M,
	XAI_PRICE_CACHED_PER_M,
	XAI_PRICE_INPUT_PER_M,
	XAI_PRICE_OUTPUT_PER_M,
} from "./config.js";

// ── CostLedger ──────────────────────────────────────────────────────────────────

interface CostBucket {
	calls: number;
	prompt_tokens: number;
	cached_tokens?: number;
	completion_tokens: number;
	cost_usd: number;
}

interface VoyageBucket {
	calls: number;
	tokens: number;
	cost_usd: number;
}

interface LedgerData {
	xai: CostBucket & { cached_tokens: number };
	google: CostBucket;
	anthropic: CostBucket;
	voyage: VoyageBucket;
	total_cost_usd: number;
	last_updated: string;
}

function freshLedger(): LedgerData {
	return {
		xai: { calls: 0, prompt_tokens: 0, cached_tokens: 0, completion_tokens: 0, cost_usd: 0 },
		google: { calls: 0, prompt_tokens: 0, completion_tokens: 0, cost_usd: 0 },
		anthropic: { calls: 0, prompt_tokens: 0, completion_tokens: 0, cost_usd: 0 },
		voyage: { calls: 0, tokens: 0, cost_usd: 0 },
		total_cost_usd: 0,
		last_updated: "",
	};
}

class CostLedger {
	private data: LedgerData = freshLedger();
	private path = "";

	async init(dataDir: string = DATA_DIR): Promise<void> {
		this.path = `${dataDir}/costs.json`;
		try {
			const file = Bun.file(this.path);
			if (await file.exists()) {
				this.data = await file.json();
			}
		} catch {
			this.data = freshLedger();
		}
	}

	private async save(): Promise<void> {
		if (!this.path) return;
		try {
			await Bun.write(this.path, JSON.stringify(this.data, null, 2));
		} catch (e) {
			console.warn("Cost ledger save error:", e);
		}
	}

	private updateTotal(): void {
		this.data.total_cost_usd = +(
			(this.data.xai?.cost_usd ?? 0) +
			(this.data.google?.cost_usd ?? 0) +
			(this.data.anthropic?.cost_usd ?? 0) +
			(this.data.voyage?.cost_usd ?? 0)
		).toFixed(8);
		this.data.last_updated = new Date().toISOString();
	}

	async recordXai(promptTokens: number, cachedTokens: number, completionTokens: number): Promise<void> {
		const cost =
			(promptTokens - cachedTokens) * XAI_PRICE_INPUT_PER_M / 1_000_000 +
			cachedTokens * XAI_PRICE_CACHED_PER_M / 1_000_000 +
			completionTokens * XAI_PRICE_OUTPUT_PER_M / 1_000_000;
		const x = this.data.xai;
		x.calls++;
		x.prompt_tokens += promptTokens;
		x.cached_tokens += cachedTokens;
		x.completion_tokens += completionTokens;
		x.cost_usd = +(x.cost_usd + cost).toFixed(8);
		this.updateTotal();
		await this.save();
	}

	async recordGoogle(promptTokens: number, completionTokens: number): Promise<void> {
		const cost =
			promptTokens * GOOGLE_PRICE_INPUT_PER_M / 1_000_000 +
			completionTokens * GOOGLE_PRICE_OUTPUT_PER_M / 1_000_000;
		if (!this.data.google) {
			this.data.google = { calls: 0, prompt_tokens: 0, completion_tokens: 0, cost_usd: 0 };
		}
		const g = this.data.google;
		g.calls++;
		g.prompt_tokens += promptTokens;
		g.completion_tokens += completionTokens;
		g.cost_usd = +(g.cost_usd + cost).toFixed(8);
		this.updateTotal();
		await this.save();
	}

	async recordAnthropic(promptTokens: number, completionTokens: number): Promise<void> {
		const cost =
			promptTokens * ANTHROPIC_PRICE_INPUT_PER_M / 1_000_000 +
			completionTokens * ANTHROPIC_PRICE_OUTPUT_PER_M / 1_000_000;
		if (!this.data.anthropic) {
			this.data.anthropic = { calls: 0, prompt_tokens: 0, completion_tokens: 0, cost_usd: 0 };
		}
		const a = this.data.anthropic;
		a.calls++;
		a.prompt_tokens += promptTokens;
		a.completion_tokens += completionTokens;
		a.cost_usd = +(a.cost_usd + cost).toFixed(8);
		this.updateTotal();
		await this.save();
	}

	async recordVoyage(tokens: number): Promise<void> {
		const cost = tokens * VOYAGE_PRICE_PER_M / 1_000_000;
		const v = this.data.voyage;
		v.calls++;
		v.tokens += tokens;
		v.cost_usd = +(v.cost_usd + cost).toFixed(8);
		this.updateTotal();
		await this.save();
	}

	snapshot(): LedgerData {
		return JSON.parse(JSON.stringify(this.data));
	}

	/**
	 * Re-read costs.json from disk — used by the dashboard server to pick up
	 * writes from the plugin process (which runs in a separate Bun instance).
	 */
	async load(): Promise<void> {
		if (!this.path) return;
		try {
			const file = Bun.file(this.path);
			if (await file.exists()) {
				this.data = await file.json();
			}
		} catch {
			// Non-fatal — keep existing data if file read fails
		}
	}

	async reset(): Promise<void> {
		this.data = freshLedger();
		this.data.last_updated = new Date().toISOString();
		await this.save();
	}
}

// ── ActivityLog ─────────────────────────────────────────────────────────────────

interface ActivityEntry {
	ts: string;
	api: string;
	model: string;
	operation: string;
	prompt_tokens: number | null;
	cached_tokens: number | null;
	completion_tokens: number | null;
	tokens: number | null;
	cost_usd: number;
}

class ActivityLog {
	private entries: ActivityEntry[] = [];
	private path = "";

	async init(dataDir: string = DATA_DIR): Promise<void> {
		this.path = `${dataDir}/activity.json`;
		try {
			const file = Bun.file(this.path);
			if (await file.exists()) {
				this.entries = await file.json();
			}
		} catch {
			this.entries = [];
		}
	}

	private async save(): Promise<void> {
		if (!this.path) return;
		try {
			await Bun.write(this.path, JSON.stringify(this.entries));
		} catch {
			// Non-fatal — activity log is ephemeral data
		}
	}

	/**
	 * Re-read activity.json from disk — used by the dashboard to pick up
	 * entries recorded by the plugin in a separate process.
	 */
	async load(): Promise<void> {
		if (!this.path) return;
		try {
			const file = Bun.file(this.path);
			if (await file.exists()) {
				this.entries = await file.json();
			}
		} catch {
			// Non-fatal — keep existing entries
		}
	}

	private append(entry: ActivityEntry): void {
		this.entries.push(entry);
		if (this.entries.length > ACTIVITY_MAX) {
			this.entries = this.entries.slice(-ACTIVITY_MAX);
		}
		// Fire-and-forget save — don't block the caller
		this.save();
	}

	recordXai(
		promptTokens: number,
		cachedTokens: number,
		completionTokens: number,
		costUsd: number,
		operation = "extraction",
	): void {
		this.append({
			ts: new Date().toISOString(),
			api: "xai",
			model: EXTRACTION_MODEL,
			operation,
			prompt_tokens: promptTokens,
			cached_tokens: cachedTokens,
			completion_tokens: completionTokens,
			tokens: null,
			cost_usd: +costUsd.toFixed(8),
		});
	}

	recordGoogle(
		promptTokens: number,
		completionTokens: number,
		costUsd: number,
		operation = "extraction",
	): void {
		this.append({
			ts: new Date().toISOString(),
			api: "google",
			model: GOOGLE_EXTRACTION_MODEL,
			operation,
			prompt_tokens: promptTokens,
			cached_tokens: null,
			completion_tokens: completionTokens,
			tokens: null,
			cost_usd: +costUsd.toFixed(8),
		});
	}

	recordAnthropic(
		promptTokens: number,
		completionTokens: number,
		costUsd: number,
		operation = "extraction",
	): void {
		this.append({
			ts: new Date().toISOString(),
			api: "anthropic",
			model: ANTHROPIC_EXTRACTION_MODEL,
			operation,
			prompt_tokens: promptTokens,
			cached_tokens: null,
			completion_tokens: completionTokens,
			tokens: null,
			cost_usd: +costUsd.toFixed(8),
		});
	}

	recordVoyage(tokens: number, costUsd: number): void {
		this.append({
			ts: new Date().toISOString(),
			api: "voyage",
			model: EMBEDDING_MODEL,
			operation: "embed",
			prompt_tokens: null,
			cached_tokens: null,
			completion_tokens: null,
			tokens,
			cost_usd: +costUsd.toFixed(8),
		});
	}

	recent(limit = 50): ActivityEntry[] {
		return [...this.entries.slice(-limit)].reverse();
	}
}

// ── Module-level singletons ─────────────────────────────────────────────────────

export const ledger = new CostLedger();
export const activityLog = new ActivityLog();
